lib/Sema/LookupVisibleDecls.cpp

```diff
@@ -822,6 +857,18 @@ class OverrideFilteringConsumer : public VisibleDeclConsumer {
     DeclsToReport.insert(FoundDeclTy(VD, Reason));
   }
 };
+
+struct ShadowedKeyPathMembers : public VisibleDeclConsumer {
+  VisibleDeclConsumer &consumer;
+  llvm::DenseSet<DeclBaseName> &seen;
+  ShadowedKeyPathMembers(VisibleDeclConsumer &consumer, llvm::DenseSet<DeclBaseName> &knownMembers) : consumer(consumer), seen(knownMembers) {}
+  void foundDecl(ValueDecl *VD, DeclVisibilityKind reason) override {
+    // Dynamic lookup members are only visible if they are not shadowed by
+    // non-dynamic members.
+    if (seen.count(VD->getBaseName()) == 0)
+      consumer.foundDecl(VD, reason);
```

Comment:
The `reason` here is necessarily accurate. Maybe, we should introduce `DeclVisibilityKind::DynamicMember`?
Also we might want to propagate `subscript` and `memberType` to the consumer, somehow 🤔

Reply:
Let's pursue this in follow-up PRs.

Reply:
I thought about a new `DeclVisibilityKind`, but hesitated because it's orthogonal from some of the existing cases:
* `MemberOfCurrentNominal`
* `MemberOfProtocolImplementedByCurrentNominal`
* `MemberOfSuper`

Ideally we could distinguish dynamic member of current nominal vs. dynamic member of super.  It's also not clear to me how to prioritize dynamic members vs regular members - does it affect priority? If so, how?

We could add three new kinds, or possibly add a separate bit, I'm not sure.

---

I agree with the overall point that we probably want to add this new information in order to handle the type mapping correctly in the future.  Do you think it would be sufficient for the lookup code to determine a type using the subscript + memberType and pass that single type back to code-completion, or do we want to bundle up all the pieces so  that code-completion can also add the contextual type before solving?